oslabs-beta/DockerLocal
# Create docker compose file from existing dockerfiles

## Description:
DockerLocal needs a way to access dockerfiles in different repositories that are stored locally to create a compose file. I have added one controller file `dockerController.ts` , a folder named `scripts`, and a shell script` findDockerfiles.sh` that work together to create a populated docker-compose.yml file that will currently appear in DockerLocal/MyRepos.

## Changes I Made:

1. ` findDockerfiles.sh`: Searches for dockerfiles in DockerLocal/MyRepos/{ProjectName} by using regex to find any **file** with docker in the name (case insensitive)
2. `dockerController.ts` : Creates template compose file with predetermined build paths and container names
                -` dockerContoller.getFilePaths`: Runs shell script to get file paths and saves an array of all paths that will be used in the build config option.
               - `dockerController.getContainerNames`: Uses Build Paths to create unique container names taken from the folder names in the build path
               - `dockerController.createDockerCompose`: Uses Build Paths and Container Names to create a template compose file.
3. `docker-route.ts`:  Middleware that uses the projectName to find all Dockerfiles and use them to create a compose file with buildpaths, ports, and container names.
4. `server.ts`: Now contains an API endpoint for creating compose file.

## How to Test:
1. npm install
2. npm start
3. Use Postman:
      1. Create a post request that sends `{"projectName": "Test"}` to localhost:3001/docker/ 
      2. Wait to receive docker-compose.yaml file
<img width="745" alt="Screen Shot 2020-07-01 at 11 13 21 PM" src="https://user-images.githubusercontent.com/32973305/86322799-781e0580-bbf0-11ea-89b4-b2755b7b700a.png">
      3. Verify docker-compose.yaml file is located in myRepos/{ProjectName}
<img width="308" alt="Screen Shot 2020-07-01 at 11 14 29 PM" src="https://user-images.githubusercontent.com/32973305/86322882-9e43a580-bbf0-11ea-863d-44d33c1f3fa2.png">

4. Contact me to delete myRepos folder with all its content before merging. Folder is **only** for __testing_ purposes_. I will also add myRepos/ to .gitignore again.
